linux/virtual_device.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2019 Hewlett Packard Enterprise Development LP
+
+package linux
+
+import (
+	"fmt"
+	"github.com/hpe-storage/common-host-libs/model"
+	"github.com/hpe-storage/common-host-libs/util"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+)
+
+const (
+	diskBySerial = "/dev/disk/by-id/"
+	devDir       = "/dev/"
+	// ScsiHostPathFormat
+	ScsiHostPathFormat = "/sys/class/scsi_host/"
+	// ScsiHostScanPathFormat
+	ScsiHostScanPathFormat = "/sys/class/scsi_host/%s/scan"
+)
+
+// GetNimbleDmDevices : get the scsi device by specified serial
+func GetDeviceBySerial(serialNumber string) (device *model.Device, err error) {
+	util.LogDebug.Printf(">>>>> GetDeviceBySerial called with serial %s", serialNumber)
+	defer util.LogDebug.Println("<<<<< GetDeviceBySerial")
+
+	files, err := ioutil.ReadDir(diskBySerial)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	for _, f := range files {
+		if !strings.Contains(f.Name(), serialNumber) {
+			continue
+		}
+		if f.Mode()&os.ModeSymlink != 0 {
+			pathName, err := os.Readlink(diskBySerial + f.Name())
+			if err != nil {
+				util.LogError.Printf("unable to read symlink %s to get device details, err %s", f.Name(), err.Error())
+				return nil, err
+			}
+			device := &model.Device{}
+			// ../../sdb
+			devNames := strings.Split(pathName, "/")
```

Comment:
nit:  check size of devNames before dereferencing it

Reply:
sure, will add this..